ob1k-cache/src/main/java/com/outbrain/ob1k/cache/metrics/MonitoringCacheDelegate.java

```diff
@@ -0,0 +1,100 @@
+package com.outbrain.ob1k.cache.metrics;
+
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Iterables;
+import com.outbrain.ob1k.cache.EntryMapper;
+import com.outbrain.ob1k.cache.TypedCache;
+import com.outbrain.ob1k.concurrent.ComposableFuture;
+import com.outbrain.ob1k.concurrent.handlers.SuccessHandler;
+import com.outbrain.swinfra.metrics.api.Counter;
+import com.outbrain.swinfra.metrics.api.MetricFactory;
+
+/**
+ * A decorator adding metric monitoring to TypedCache.
+ *
+ * @author hunchback
+ */
+public class MonitoringCacheDelegate<K, V> implements TypedCache<K, V> {
+  protected final TypedCache<K, V> delegate;
+
+  /* interface call metrics */
+  private final AsyncOperationMetrics<V> getAsyncMetrics;
+  private final AsyncOperationMetrics<Map<K, V>> getBulkAsyncMetrics;
+  private final AsyncOperationMetrics<Boolean> setAsyncMetrics;
+  private final AsyncOperationMetrics<Map<K, Boolean>> setBulkAsyncMetrics;
+  private final AsyncOperationMetrics<Boolean> deleteAsyncMetrics;
+
+  /* cache metrics */
+  private final Counter total;
+  private final Counter hits;
+
+  private final SuccessHandler<V, V> getCacheMetricsUpdater = new SuccessHandler<V, V>() {
+    @Override
+    public V handle(final V result) throws ExecutionException {
+      total.inc();
+      if (result != null) {
+        hits.inc();
+      }
+      return result;
+    }
+  };
+
+  public MonitoringCacheDelegate(final TypedCache<K, V> delegate, String cacheName, final MetricFactory metricFactory) {
+    Preconditions.checkNotNull(delegate);
```

Comment:
Preconditions returns the tested object, so
this.delegate = Preconditions.checkNotNull(delegate);